### ISSUE TITLE
Memory leak in LXEntryFormatter

### ISSUE THREAD

**Issue by rhcpfan** [logkit/logkit#26]
Hello,

I have noticed that every time I call a method to log some information (log.error(), log.info(), log.notice(), etc.) I see a memory leak in the Instruments tool (using Leaker)

I have a global instance of the logger in my `AppDelegate.swif` file, declared as follows:

``` swift
import LogKit
let logger = createLogger(.All, fileExtension: "txt")
```

The definition of the `createLogger` function is the following:

``` swift
func createLogger (minimumPriorityLevel: LXPriorityLevel, fileExtension: String) -> LXLogger {
    // Build the logFilePath by using the current date
    // ...
    //
    let logFileEndpoint = LXFileEndpoint(fileURL: logFilePath, 
                                         shouldAppend: true, 
                                         minimumPriorityLevel: minimumPriorityLevel, 
                                         dateFormatter: LXDateFormatter.standardFormatter(),
                                         entryFormatter: LXEntryFormatter.standardFormatter())
    return LXLogger(endpoints: [LXConsoleEndpoint(), logFileEndpoint])
}
```

In the `viewDidLoad()` method of the first viewController I call `logger.notice("View has loaded")`
I have a button that calls `logger.info("Button has been pressed")`

When I start the profiling, the first screenshot detects the memory leak from the first `logger.notice()` call. The next screenshots says that we have "No new Leaks". After I press the button and take another screenshot, it says "2 new leaks"

Below is a screen capture of the call tree shown by Instruments

![screen shot 2016-06-06 at 12 31 07](https://cloud.githubusercontent.com/assets/3796970/15817900/f763ea50-2be2-11e6-8907-f3f54964f9df.png)

And this is the last method in the tree that I can see:
![screen shot 2016-06-06 at 12 34 44](https://cloud.githubusercontent.com/assets/3796970/15817918/142fe4ea-2be3-11e6-99be-8287dbb84a04.png)

**Comment by rhcpfan** [comment]
Hello again,

I have noticed that if I'd use **string concatenation** in the `standardFormatter()` method, there are no leaks.

I have changed the `standardFormatter()` method to use string concatenation as follows:

``` swift
/// Converts `LXLogEntry` objects into strings in a standard format that contains basic debugging information.
public class func standardFormatter() -> Self { return self.init { e in

    // Append date-time
    var entryString = e.dateTime
    // Append message type
    entryString += " [" + e.level.uppercaseString + "] "
    // Append function name
    entryString += e.functionName
    // Append file name
    entryString += " <" + e.fileName + "> "
    // Append the message
    entryString += e.message

    return entryString }
}
```

With this implementation I get no memory leaks:

![screen shot 2016-06-06 at 15 52 48](https://cloud.githubusercontent.com/assets/3796970/15822501/c0fa8a84-2bfe-11e6-85a2-a2b6271ef214.png)

I hope this would help you in some way!

Best regards,
**rhcpfan**

**Comment by justinpawela** [comment]
Hi @rhcpfan,

Thanks much for reporting this discovery. I'm sorry it has taken so long for me to respond.

I was able to replicate your memory leak, and to confirm that your string concatenation trick fixed the leak. After some searching, I found that this is an [open bug](https://bugs.swift.org/browse/SR-1728) in the Swift Standard Library. There also seems to be a [related radar](https://openradar.appspot.com/26366199).

After some trail and error I have settled on using `String(format:_:)` in the predefined formatters to serialize Log Entries. I will soon release an update containing these fixes.

Thanks again!
